Package for Python 3 on Fedora

kahowell opened the pull request:
#1743 should land first, and then I'll just change the base to master. I wanted to go ahead and get this out for folks to review early (if they'd like).

kahowell commented:
ab828647..f1cb3784: merge master, handle `DESTDIR`

jirihnidek commented on subscription-manager.spec:
```diff
@@ -33,6 +33,14 @@
+%global __python %__python3
+%global py_package_prefix python3
+%else
+%global py_package_prefix python
```
What about RHEL8? Do we know about convention used there?

jirihnidek commented on subscription-manager.spec:
```diff
@@ -113,16 +121,23 @@ BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
+Requires:  %{py_package_prefix}-decorator
 Requires:  virt-what
+%if %use_python3
+Requires:  python3-subscription-manager-rhsm = %{version}
```
Why is different convection used? Why is not susbscription-manager-rhsm used too? Is it due to virt-who?

kahowell replied:
Actually, I was trying to follow the Fedora packaging guidelines: https://fedoraproject.org/wiki/Packaging:Naming?rd=Packaging:NamingGuidelines#Python3_binary_package_naming

jirihnidek replied:
OK. I didn't know about this rule.

jirihnidek commented on subscription-manager.spec:
```diff
@@ -259,7 +278,11 @@ Summary: Subscription Manager plugins for DNF
 Requires: %{name} = %{version}-%{release}
 Requires: dnf >= 1.0.0
+%if %use_python3
+Requires: python3-dnf-plugins-core
```
I guess that we can use `%{py_package_prefix}` here too.

jirihnidek commented on subscription-manager.spec:
```diff
@@ -333,6 +357,7 @@ Obsoletes: python-rhsm <= 1.20.3-1
 A small library for communicating with the REST interface of a Red Hat Unified
 Entitlement Platform. This interface is used for the management of system
 entitlements, certificates, and access to content.
+%endif
```
The e.g. description of package subscription-manager-rhsm is there twice. The logic could be IMHO simplified for this package.

kahowell commented:
> BTW: do we want to use python3 as default in all versions of fedora? (line 10)

In all supported versions of Fedora (26 and 27 at this time), `dnf` uses `python3`.

Actually, it might be more consistent to just use the logic dnf uses verbatim... I'll update this PR to do that.

https://github.com/rpm-software-management/dnf/blob/dde4683b7167c4166603c4e0b3eb0c00bf017db1/dnf.spec#L22-L26

kahowell commented:
@jirihnidek I addressed some of your concerns and also refactored to use the python3 prefix macros for EPEL (see https://fedoraproject.org/wiki/PackagingDrafts:Python3EPEL), as that seems more compatible. I'm still not entirely convinced we have the python-rhsm python3 package naming correct... (I think it should be python3-rhsm based on the Fedora package naming guidelines...), but maybe we're a special case since it's a component of an application and less of a library for general usage? @awood thoughts?

awood commented:
> I think it should be python3-rhsm based on the Fedora package naming guidelines...

Yeah, I think they want it as `python2-rhsm` and `python3-rhsm`.  Someone in Fedora actually made some changes to that effect already and broke some of the dependencies.

kahowell commented:
- updated python-rhsm naming to be python-rhsm or python3-rhsm depending on which is used
- updated to fix SUSE builds

I've asked on Fedora devel email list about python package naming guidelines (and whether we should o anything else while we're tweaking the specfile), will comment back the gist of what I get back if they don't end up as GitHub comments.

awood commented on subscription-manager.spec:
```diff
@@ -113,16 +123,19 @@ BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
-Requires:  python-decorator
-Requires:  python-six
-Requires:  python-dateutil
+Requires:  %{py_package_prefix}-rhsm = %{version}
```
I thought python-rhsm went away and we were using subscription-manager-rhsm now?

kahowell replied:
Yeah, that is what we were doing as of #1712... but I was under the impression that we actually are required to call it python-rhsm or python3-rhsm by the fedora package guidelines?

awood replied:
Oh, sorry.  When you asked earlier I had forgotten that we had moved away from python-rhsm.  I actually don't know what they would want.  I mean, I suppose they would want the whole thing to be python-subscription-manager or python3-subscription-manager.  I suppose we should ask someone?  Since we've dumped python-rhsm, my vote is just to stick with subscription-manager-rhsm instead of prefixing it.

kahowell replied:
@awood @jirihnidek I haven't gotten much feedback from Fedora devel lists. @xsuchy did give me some suggestions though; thanks Miroslav! I think what I'll do for now then is this: I'll revert the naming changes from this PR and handle any naming changes in a separate PR. Then this PR will be purely for getting the packages built in Fedora using Python 3.

kahowell commented:
FYI, pushed some changes, especially not touching the names in this PR.

kahowell commented:
Also, to be handled in a separate PR: we don't want to support the GUI in Python 3... so we'll stop building/packaging the GUI when using Python 3...

kahowell commented:
@awood @jirihnidek any other feedback? I think I've addressed everything (all the things I haven't deferred to another PR anyways).

awood commented:
@kahowell Looks good to me.  Nicely done. 💎 🎆
